### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by romalytvynenko**
Initial implementation of paged component

### CODE REVIEW

**Comment by XedinUnknown** [comment]
It's true that we cannot assume any particular purpose of this component; only that it has 0 or more pages, and that it knows about them. However, I don't really understand right now what purpose this component serves, especially since the bulk of its functionality seems to be in the root template. I cannot see usage of the `collection` prop anywhere.

Also, I thought that this component would somehow extend the `Repeater` component. Each page has something in common, and something different. Then, the Paged component repeats these pages, rendering them and their content. Extensions can add on top of that to make tabs, or a slider. The visible/active pages are controlled by the limited collection. That's what I had in mind.

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            /**
+             * As far in Vue we need to wrap component's template
+             * in some root element, consumer of `Paged` can pass his
+             * wrapper for pages.
+             *
+             * @type {Function} Wrapper for the page root
+             */
+            pageWrapper: {
+                default () {
+                    return (pageTemplate, page = null) => {
```
Comment (XedinUnknown) [inline src/index.js]: The `page` argument seems to be unused. What is the role of this argument?

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        },
+
+        methods: {
+            /**
+             * Check whether a specific page is active.
+             *
+             * @param {Page} item Some specific `Page` to be checked
+             * @return {Boolean} Is page active
+             */
+            isPageActive (item) {
```
Comment (XedinUnknown) [inline src/index.js:74]: Missing `since` tag.

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+             * @type {LimitedCollection} Collection of active pages keys.
+             */
+            selectedKeys: {
+                required: true
+            },
+
+            /**
+             * @type {string} `Paged` wrapper.
+             */
+            wrapper: {
```
Comment (XedinUnknown) [inline src/index.js]: Missing `since` tag.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    return Vue.extend({
+        props: {
+            items: {
+                required: true
+            },
+
+            /**
+             * @type {LimitedCollection} Collection of active pages keys.
+             */
+            selectedKeys: {
```
Comment (XedinUnknown) [inline src/index.js]: Missing `since` tag.

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
+export function CfPaged (Vue) {
+    return Vue.extend({
+        props: {
+            items: {
```
Comment (XedinUnknown) [inline src/index.js]: Missing `since` tag.

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+export function CfPaged (Vue) {
```
Comment (XedinUnknown) [inline src/index.js:11]: Missing `since` tag.

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            },
+
+            /**
+             * As far in Vue we need to wrap component's template
+             * in some root element, consumer of `Paged` can pass his
+             * wrapper for pages.
+             *
+             * @type {Function} Wrapper for the page root
+             */
+            pageWrapper: {
```
Comment (XedinUnknown) [inline src/index.js]: Missing `since` tag.

File: src/index.js (hunk 2 above)
Comment (XedinUnknown) [inline src/index.js:74]: Missing `since` tag.

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+             *
+             * @param {Page} item Some specific `Page` to be checked
+             * @return {Boolean} Is page active
+             */
+            isPageActive (item) {
+                return this.selectedKeys.hasItem(item.id)
+            }
+        },
+
+        render (h) {
```
Comment (XedinUnknown) [inline src/index.js]: Missing docs.

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                let ctx = {
+                    items: instance.items,
+                    isActive: instance.isPageActive(page)
+                }
+
+                let component = Vue.extend({
+                    data () {
+                        return ctx
+                    },
+                    template: this.pageWrapper(page.render(ctx), page)
```
Comment (XedinUnknown) [inline src/index.js:102]: How does the data of the page itself get passed to the template?

File: tests/e2e/app.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+Vue.component('avatar', {
+    props: ['name'],
+    render(h) {
+        return h('div', [
+            h('h2', [this.name])
+        ])
+    }
+});
+
+document.addEventListener('DOMContentLoaded', () => {
```
Comment (XedinUnknown) [inline tests/e2e/app.js]: Looks like this test cannot be run in a non-browser environment. How can it be changed to work with a simulated DOM?

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
- Documentation and tags missing, including for tests.
- Not sure how template gets data of the current subject.
- Tests don't seem to be usable outside of the browser, and use a global.

**Comment by romalytvynenko** [comment]
Pushed that changes. 

About running that test in non-browser environment. Should we run Vue component's e2e test on non-browser environment? This is strange for me, because while looking for framework I saw that:
1. All JS e2e frameworks work inside browser environments. 
2. e2e testing point is to test something in it's environment, so Vue component's e2e test should be run in browser. Or I've missed something?

**Comment by XedinUnknown** [comment]
The browser, document, window etc are not the concern of template rendering. Rendering is related to the DOM. If you can have the DOM, but not everything else, this would be the perfect test.

**Comment by romalytvynenko** [comment]
@XedinUnknown, I've removed `document` dependency

**Review by XedinUnknown — APPROVED** [review]

**Pull request merged**